feat(backup): aws backup for s3 buckets

PeterBaker0 opened the pull request:
This is my first contribution to the wonderful CDK - so apologies if I've missed something. 

This PR is a non breaking change which adds support for S3 buckets to the AWS Backup L2 constructs. 

This was realised by

- a `fromS3Bucket` static method on the `BackupResource` construct
- a check for the bucket type in the Backupable Resources Collector Aspect

When using this construct for a project, I found two other issues 

1. The `BackupSelection` construct does not add the S3 backup/restore managed policies and has no means to do so
2. The `BackupSelection` construct does not expose the IAM role (at least directly - it was previously set as the `grantPrincipal`)

So this PR also:

1. Adds `allowS3Backup` and `allowS3Restores` flags which add the AWS Backup for S3 managed policies for backup/restore respectively 
2. Exposes the `role` in the BackupSelection construct

Additionally, I found that in using this construct, it was not possible to provide a fully customised role to the construct, it would always add, at least, the managed policy for backup. So I added a `customRole` flag which can be provided to disable the addition of any managed policies, meaning that a user could provide a completely custom role, if desired. Annotation warnings for the construct were added in the case where both the `customRole` flag and any of the `allowX` flags for managed policies were both `true`.

An integration test was added which makes sure that selections both using the `fromS3Bucket` and `fromConstruct` resource static methods function. The `allowS3Backup` and `allowS3Restore` flags were true to ensure that the managed policy names were correct. 

Added unit tests for the S3 support (static resource methods, managed policies), the IAM role property, and the `customRole` 
flag behaviour. 

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: cfaaca2bf7115e1c4a4cbde35577fa57629d6a4a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e75dbf94-9aca-4a57-94f4-9207933cbc41%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

Naumel commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -69,6 +71,22 @@ plan.addSelection('Selection', {
+const iamRole = selection.role;
+
+// add example fake managed policy to auto generated IAM role
+const fakePolicyName = 'FakePolicyName';
```
`Fake` sounds harsh, why not `Managed`?

PeterBaker0 replied:
Sure :+1:

Naumel commented on packages/@aws-cdk/aws-backup/README.md:
````diff
@@ -180,7 +198,7 @@ const vault = new backup.BackupVault(this, 'Vault', {
 ```
 
-Alternativately statements can be added to the vault policy using `addToAccessPolicy()`.
+Alternatively, statements can be added to the vault policy using `addToAccessPolicy()`.
````
😄 good catch

Naumel commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -218,3 +236,27 @@ const role = new iam.Role(this, 'Access Role', { assumedBy: new iam.ServicePrinc
+
+## Backup selection IAM role policies
+
+When creating a new `BackupSelection`, if no user IAM role is provided, a new
```
Why `user` in the sentence? Can it not be skipped completely?

PeterBaker0 replied:
Good catch

Naumel commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -218,3 +236,27 @@ const role = new iam.Role(this, 'Access Role', { assumedBy: new iam.ServicePrinc
+When creating a new `BackupSelection`, if no user IAM role is provided, a new
+IAM role with a managed policy for backup will be created for the selection. 
+
+All policies, by default, include the `AWSBackupServiceRolePolicyForBackup` AWS
```
An extra reference never hurts, e.g. https://docs.aws.amazon.com/aws-backup/latest/devguide/iam-service-roles.html

PeterBaker0 replied:
Okay - will add in - thanks

Naumel commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -218,3 +236,27 @@ const role = new iam.Role(this, 'Access Role', { assumedBy: new iam.ServicePrinc
+`AWSBackupServiceRolePolicyForS3Backup` managed policy). To enable S3 restores
+for a selection, include the `allowS3Restores` flag (the
+`AWSBackupServiceRolePolicyForS3Restore` managed policy). Both of these flags
+are `false` by default.
```
👌

Naumel commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -218,3 +236,27 @@ const role = new iam.Role(this, 'Access Role', { assumedBy: new iam.ServicePrinc
+`AWSBackupServiceRolePolicyForS3Restore` managed policy). Both of these flags
+are `false` by default.
+
+If the generated IAM roles for a `BackupSelection` are insufficient, you can
+either include your own IAM role using the optional `role` field, or access the
+generated IAM role with the read only `role` property. If you want to have
+complete control over role policies, you should provide your own role with the
+`role` field and disable any automatic addition of managed policies with the
+`customRole` field.
```
This is just me being extra, but if you find it in your heart to add the code samples going from one option to the other, that's be 🙌

PeterBaker0 replied:
Happy to do that - not extra at all :rofl:

Naumel commented on packages/@aws-cdk/aws-backup/lib/selection.ts:
```diff
@@ -85,12 +118,40 @@ export class BackupSelection extends Resource implements iam.IGrantable {
+        role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AWSBackupServiceRolePolicyForS3Restore'));
+      }
+    } else {
+      // Provide deployment warnings if customRole is enabled but any of the
+      // permission flags are true
+      if (props.allowRestores) {
+        Annotations.of(this).addWarning('Both the customRole and allowRestores flag was set to true - no restore managed policy was added.');
+      }
+      if (props.allowS3Backup) {
+        Annotations.of(this).addWarning('Both the customRole and allowS3Backup flag was set to true - no S3 backup managed policy was added.');
+      }
+      if (props.allowS3Restores) {
+        Annotations.of(this).addWarning('Both the customRole and allowS3Restores flag was set to true - no S3 restore managed policy was added.');
+      }
```
Should the customer role not be set somewhere though, as per
>So I added a customRole flag which can be provided to disable the addition of any managed policies, meaning that a user could provide a completely custom role, if desired.

PeterBaker0 replied:
Not sure what you mean here - I think maybe `customRole` is not a good name for the field. 

Personally I think the default behaviour of the service role policy being added even when the user provides their own Role is debatable/opinionated - this flag is to add the capacity to have the construct just reference your provided role or create one on your behalf without adding any policies to it. 

Happy for ideas on how to clarify this - but as I said I don't fully understand your comment above.

PeterBaker0 replied:
Maybe another approach would be to add an `allowBackup` flag (similar to `allowRestore`) which is default to `true`. Then all of the managed policies can just be switched on/off, but a reasonable default selection of policies is present. This also means there would be no breaking change to the API.

Naumel reviewed (dismissed):
Thank you for taking the time and submitting a PR.
Most things look good and this is a nice addition.

I have a number for positive remarks and questions scattered through your change. 

The most pressing one I will write here, please add unit tests, especially since there are misconfiguration cases.

👍 thank you for the description too, really makes contextualization easy.

PeterBaker0 commented:
> The most pressing one I will write here, please add unit tests, especially since there are misconfiguration cases.

I did add a few unit tests in the selections test - this covered the `fromS3Bucket`, `fromConstruct`, as well as the changes to the roles. Wasn't sure how to handle the unit test case for misconfiguration - how do you enforce that there should be annotation warnings?

PeterBaker0 commented:
@Naumel thanks for your review. 

I think the `customRole` interface was confusing so changed it to include a `allowBackup` flag which defaults to `true`. This means that there is no breaking change, however users can easily specify `allowBackup` as `false` if they need to use a custom role which doesn't include this managed policy. 

As mentioned above - I did add a few unit tests and update relevant existing ones - could you clarify which features are missing from the unit test coverage if you would like more? Thanks. 

I also added another new integration test to verify that the various combinations of policies are all valid and deploy properly. 

Readme was updated and streamlined. Added a table to map the allowX fields to the managed policies, and supplied a link to relevant section in AWS docs. 

Cheers.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: e3fbb1c30d3668e4b65890c14847bac1803f2c83
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fe22c55b-3b3b-4298-95fe-b9ab8284f476%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: bd6f4d5c7e43d616b90cae33f6b42c6087871026
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d84030f1-0754-4e8e-a791-be3ad1870051%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

Naumel reviewed (approved):
I like the renaming, makes reading easier for sure.

If you decide to add even more functionality, I look forward to the submission(s)!

Naumel commented:
@Mergifyio update

PeterBaker0 commented:
> I like the renaming, makes reading easier for sure.
> 
> If you decide to add even more functionality, I look forward to the submission(s)!

Thanks! 

There was a stale reference in the parameter comments so have corrected that now - apologies if that breaks your PR approval!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4ed45a4e4a050d50dd86e3007560e59c93c26309
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=75a9ae5d-2569-4f7b-89c4-543bd81d24a7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: e850a0b4fd6360539b45bb36d3320b07954abbd4
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fc4caad7-670b-4282-a860-111526486c47%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

PeterBaker0 commented:
@Naumel I have read the contribution guide but couldn't find any advice on the next steps. Do I need to take any more actions to get this merged or will happen automatically? Cheers.

kaizencc commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -60,14 +62,17 @@ plan.addSelection('Selection', {
-created for the selection. The `BackupSelection` implements `IGrantable`.
+created for the selection. The `BackupSelection` implements `IGrantable`. For
+more information about how to customize the IAM role used by the
+`BackupSelection`, see the _Backup selection IAM role policies_ section.
```
```suggestion
`BackupSelection`, see the [Backup selection IAM role policies](#backup-selection-iam-role-policies) section.
```

kaizencc commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -36,6 +37,7 @@ Assigning resources to a plan can be done with `addSelection()`:
 declare const plan: backup.BackupPlan;
 declare const vpc: ec2.Vpc;
 const myTable = dynamodb.Table.fromTableName(this, 'Table', 'myTableName');
+const myBucket = new s3.Bucket(this, 'Bucket', {});
```
```suggestion
const myBucket = new s3.Bucket(this, 'Bucket');
```

kaizencc commented on packages/@aws-cdk/aws-backup/README.md:
```diff
@@ -218,3 +223,74 @@ const role = new iam.Role(this, 'Access Role', { assumedBy: new iam.ServicePrinc
+  allowRestores: true,
+
+  // S3 specific policies
+  allowS3Backup: true,
+  allowS3Restores: true,
```
Question: would someone ever send in a s3 bucket in `resources` and not want to set `allowS3Backup`? To me, it seems like they go together and if so, we should find some way to determine if a resource is an s3 bucket and automatically add the s3 specific policy.

Same goes for `allowS3Restores`, I feel like we should be able to bake it into `allowRestores` and provide a better user experience.

kaizencc reviewed (commented):
Hi @PeterBaker0! To answer your question, normally once you have approval you're good to go and a maintainer will figure out how to wrestle with mergify to get your PR in. However, I chanced on this PR after reading the corresponding issues, and I have a few questions reading through this PR. Hope you can indulge me.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ff8cf506c96aca4b58be8e6c79caa82b83d93fa6
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a58594c7-bf9d-4c3e-b3a0-c52100683eda%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
@Mergifyio unqueue

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ea8e49ce960debfa2e4bc205b6fb509ef12667e6
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d09c0e17-b89f-42de-aa79-a96d9979e3b6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
@Mergifyio update

TheRealAmazonKendra reviewed (changes requested):
Putting this back into changes requested given @kaizencc's comments.

TheRealAmazonKendra commented:
One quick question from me: is there an issue for this feature? If so, please link it so it will get closed when this is merged. If not, no worries.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 46448f53a934bfd080ac90f537e471edb28494a6
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=696802e6-ffb3-4964-9b85-d7ee1f986356%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

kaizencc commented:
@PeterBaker0 are you still working on this PR? I see some 👍 on my previous comments but no action besides that. If you need more time, that's fine! If you need to drop and let someone else pick this up, that's fine too. Just let me know.

PeterBaker0 commented:
> @PeterBaker0 are you still working on this PR? I see some +1 on my previous comments but no action besides that. If you need more time, that's fine! If you need to drop and let someone else pick this up, that's fine too. Just let me know.

Hey Kaizen - I'd like to do this but I don't see myself having time in the next few weeks. 

Not sure of the process to defer this - I'm happy to close this issue then maybe we can make an issue to track the progress so far and what was missing from this PR, so someone could come and pick it up (or me if I get time)?

corymhall commented:
> Not sure of the process to defer this - I'm happy to close this PR then maybe we can make an issue to track the progress so far and what was missing from this PR, so someone could come and pick it up (or me if I get time)?

Thanks for letting us know @PeterBaker0! I'll go ahead and close this issue for now. If you ever want to come back to it you can just re-open this PR.

johnmarcou commented:
@corymhall There is any in progress PR on this subject?

Thank you!

PeterBaker0 commented:
@johnmarcou This was closed as the PR required some more in depth changes which I didn't have time to continue. 

I am unsure if someone has picked up or is working on this. 

You're welcome to continue from where I got to if you wish.
